File: Projeto Didio Pizza/Didio Pizza Springboot/backend/src/main/java/com/pizzaria/didiopizza/service/CarrinhoService.java
package com.pizzaria.didiopizza.service;

import java.util.List;
import java.util.Map;

import org.springframework.stereotype.Service;

import com.pizzaria.didiopizza.repository.CarrinhoRepository;

@Service
public class CarrinhoService {
    private final CarrinhoRepository repo;

    public CarrinhoService(CarrinhoRepository repo) {
        this.repo = repo;
    }

    public void adicionarAoCarrinho(Integer clienteId, Integer produtoId, Integer qtd) {
        if (qtd == null || qtd <= 0)
            throw new RuntimeException("Quantidade deve ser maior que zero.");
        var carrinho = repo.buscarCarrinhoAtivo(clienteId);
        Integer carrinhoId = carrinho == null ? repo.criarCarrinho(clienteId)
                : ((Number) carrinho.get("id")).intValue();
        var item = repo.buscarItem(carrinhoId, produtoId);
        if (item != null)
            repo.atualizarQuantidade(((Number) item.get("id")).intValue(), qtd);
        else
            repo.adicionarItem(carrinhoId, produtoId, qtd);
    }

    public List<Map<String, Object>> verCarrinho(Integer clienteId) {
        var c = repo.buscarCarrinhoAtivo(clienteId);
        return c == null ? List.of() : repo.listarItens(((Number) c.get("id")).intValue());
    }

    public int removerDoCarrinho(Integer clienteId, Integer produtoId) {
        var c = repo.buscarCarrinhoAtivo(clienteId);
        return c == null ? 0 : repo.removerItem(((Number) c.get("id")).intValue(), produtoId);
    }
}

File: Projeto Didio Pizza/Didio Pizza Node/backend/src/service/carrinhoService.js
import * as carrinhoRepository from '../repository/carrinhoRepository.js';


export async function adicionarAoCarrinho(clienteId, produtoId, quantidade) {

    if (quantidade <= 0) {
        throw new Error('Quantidade deve ser maior que zero.');
    }

    let carrinho = await carrinhoRepository.buscarCarrinhoAtivo(clienteId);

    if (!carrinho) {
        let carrinhoId = await carrinhoRepository.criarCarrinho(clienteId);

        carrinho = {
            id: carrinhoId
        };
    }

    let item = await carrinhoRepository.buscarItem(carrinho.id, produtoId);

    if (item) {
        await carrinhoRepository.atualizarQuantidade(item.id, quantidade);
    } else {
        await carrinhoRepository.adicionarItem(carrinho.id, produtoId, quantidade);
    }
}


export async function verCarrinho(clienteId) {

    let carrinho = await carrinhoRepository.buscarCarrinhoAtivo(clienteId);

    if (!carrinho) {
        return [];
    }

    let itens = await carrinhoRepository.listarItens(carrinho.id);

    return itens;
}


export async function removerDoCarrinho(clienteId, produtoId) {

    let carrinho = await carrinhoRepository.buscarCarrinhoAtivo(clienteId);

    if (!carrinho) {
        return 0;
    }

    let linhas = await carrinhoRepository.removerItem(carrinho.id, produtoId);

    return linhas;
}